File: java/day11_part1_2015.java

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class solution {

    public static void main(String[] args) {
        String currentPassword = readInput("input.txt");
        String newPassword = findNextPassword(currentPassword);
        System.out.println(newPassword);
    }

    public static String readInput(String filename) {
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            return br.readLine();
        } catch (IOException e) {
            return "";
        }
    }

    public static String findNextPassword(String password) {
        while (true) {
            password = incrementPassword(password);
            if (isValidPassword(password)) {
                break;
            }
        }
        return password;
    }

    public static String incrementPassword(String password) {
        char[] chars = password.toCharArray();
        for (int i = chars.length - 1; i >= 0; i--) {
            chars[i]++;
            if (chars[i] > 'z') {
                chars[i] = 'a';
            } else {
                break;
            }
        }
        return new String(chars);
    }

    public static boolean isValidPassword(String password) {
        return hasStraight(password) && !containsInvalidLetters(password) && hasTwoPairs(password);
    }

    public static boolean hasStraight(String password) {
        for (int i = 0; i < password.length() - 2; i++) {
            if (password.charAt(i) + 1 == password.charAt(i + 1) && password.charAt(i) + 2 == password.charAt(i + 2)) {
                return true;
            }
        }
        return false;
    }

    public static boolean containsInvalidLetters(String password) {
        for (char c : password.toCharArray()) {
            if (c == 'i' || c == 'o' || c == 'l') {
                return true;
            }
        }
        return false;
    }

    public static boolean hasTwoPairs(String password) {
        int count = 0;
        for (int i = 0; i < password.length() - 1; i++) {
            if (password.charAt(i) == password.charAt(i + 1)) {
                count++;
                i++; // Skip the next character
            }
        }
        return count >= 2;
    }
}

File: csharp/day11_part1_2015.cs

using System;
using System.IO;
using System.Linq;

class Program
{
    static void Main()
    {
        string currentPassword = File.ReadAllText("input.txt").Trim();
        string newPassword = FindNextPassword(currentPassword);
        Console.WriteLine(newPassword);
    }

    static string FindNextPassword(string password)
    {
        do
        {
            password = IncrementPassword(password);
        } while (!IsValidPassword(password));
        return password;
    }

    static string IncrementPassword(string password)
    {
        char[] chars = password.ToCharArray();
        for (int i = chars.Length - 1; i >= 0; i--)
        {
            chars[i]++;
            if (chars[i] > 'z')
            {
                chars[i] = 'a';
            }
            else
            {
                break;
            }
        }
        return new string(chars);
    }

    static bool IsValidPassword(string password)
    {
        return HasStraight(password) && !ContainsInvalidLetters(password) && HasTwoPairs(password);
    }

    static bool HasStraight(string password)
    {
        for (int i = 0; i < password.Length - 2; i++)
        {
            if (password[i] + 1 == password[i + 1] && password[i] + 2 == password[i + 2])
            {
                return true;
            }
        }
        return false;
    }

    static bool ContainsInvalidLetters(string password)
    {
        return password.Contains('i') || password.Contains('o') || password.Contains('l');
    }

    static bool HasTwoPairs(string password)
    {
        int count = 0;
        for (int i = 0; i < password.Length - 1; i++)
        {
            if (password[i] == password[i + 1])
            {
                count++;
                i++; // Skip the next character
            }
        }
        return count >= 2;
    }
}
